server/src/main/java/org/elasticsearch/common/logging/HeaderWarning.java

```diff
@@ -32,28 +31,58 @@
  * The result will be returned as HTTP response headers.
  */
 public class HeaderWarning {
+
+    private static final String semanticVersionPattern = "\\d+\\.\\d+\\.\\d+(?:-(?:alpha|beta|rc)\\d+)?(?:-SNAPSHOT)?";
+
+    private static final boolean hasSemanticVersion = Pattern.matches(semanticVersionPattern, Build.current().version());
```

Comment:
The change here tries to keep the current format close to the one today, checking if `version()` contains a semantic version or just an opaque string (e.g. in serverless). Let me know if it makes sense, or if we should just remove version from the warn agent, or if we should relax the check and allow basically any string for the "version" part of the agent.

Reply:
In principle I'm a little worried about the change here. I'm not sure that we have a good way to unit-test the non-semantic code path within the Elasticsearch code base.

I think I'd be happy with it if we could make the `getPatternWithSemanticVersion` and `getPatternWithoutSemanticVersion` methods visible for testing, and adjust our unit tests to cover both cases.

I don't have any objection to removing the semantic version check part of the regex altogether, but we should see if @pgomulka has concerns about that.

Reply:
Adding unit tests was definitely my plan, sorry for not mentioning it: I wanted to see if you liked this or not, or if it was better to skip (match all) the semantic version part from the regex. Let me know, I will add tests accordingly

Reply:
I think it looks good. Let's just make sure that at least semantic path is tested in elasticsearch 'statefull' repo. The opaque string we can test in serverless repo (if needed)

I am not sure what is the future of deprecations in serverless. It wasn't decided yet.

Reply:
Thanks for the input, I will keep the change and add tests then

Reply:
I've added a couple of tests here, and prepared a PR for serverless to merge once this is updated/brought into the serveless repo: https://github.com/elastic/elasticsearch-serverless/pull/952 

If we don't like it and we want to be able to run all tests in the ES codebase, we can think about alternatives, finding a way to inject `Build` (via internal methods or via  `BuildExtension` - even if this might be elaborate, see `ExtensionLoaderTests`).
Let me know!

Reply:
I think the unit tests in serverless look good. I think it'd be good to put a comment in this codebase along the lines of `// Build.current().version() may not be semantic in downstream projects or future releases` so that no one mistakenly sees this code as needing cleanup.

Reply:
Very good point, will add it